sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala

```diff
@@ -51,18 +58,209 @@ object QueryExecutionErrors {
       s"[BUG] logical plan should not have output of char/varchar type: $leaf")
   }
 
-  def cannotEvaluateGeneratorError(generator: UnresolvedGenerator): Throwable = {
-    new UnsupportedOperationException(s"Cannot evaluate expression: $generator")
+  def cannotEvaluateExpressionError(expression: Expression): Throwable = {
+    new UnsupportedOperationException(s"Cannot evaluate expression: $expression")
   }
 
-  def cannotGenerateCodeForGeneratorError(generator: UnresolvedGenerator): Throwable = {
-    new UnsupportedOperationException(s"Cannot generate code for expression: $generator")
+  def cannotGenerateCodeForExpressionError(expression: Expression): Throwable = {
+    new UnsupportedOperationException(s"Cannot generate code for expression: $expression")
   }
 
   def cannotTerminateGeneratorError(generator: UnresolvedGenerator): Throwable = {
     new UnsupportedOperationException(s"Cannot terminate expression: $generator")
   }
 
+  def castDownLongCauseOverflowError(t: Any, targetType: String): Throwable = {
+    new ArithmeticException(s"Casting $t to $targetType causes overflow")
+  }
+
+  def cannotRepresentDecimalError(value: Decimal, decimalType: DecimalType): Throwable = {
+    new ArithmeticException(s"${value.toDebugString} cannot be represented as " +
+      s"Decimal(${decimalType.precision}, ${decimalType.scale}).")
```

Comment:
It seems `Decimal` has the same error handling, so how about sharing it between here and `Decimal` just like this?
```
object Decimal {
  def some_name(...) {...}
}

def cannotRepresentDecimalError(value: Decimal, decimalType: DecimalType): Throwable = {
  Decimal.some_name(value, decimalType)
}

```
https://github.com/apache/spark/blob/8e11ce5378a2cf69ec87501e86f7ed5963649cbf/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala#L368-L369

Reply:
Maybe, the `ArithmeticException` in Decimal will be unified into file `QueryExecutionErrors` in the future.